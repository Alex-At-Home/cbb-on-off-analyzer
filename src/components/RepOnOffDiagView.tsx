// React imports:
import React, { useState } from 'react';

// Lodash
import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Library Utils
// @ts-ignore
import { mean, std } from "mathjs";

// Utils
import { OnOffReportDiagUtils } from "../utils/OnOffReportDiagUtils";

type Props = {
  diagInfo: Array<any>,
  player: Record<string, any>,
  expandedMode: boolean,
  showHelp: boolean
};

const RepOnOffDiagView: React.FunctionComponent<Props> = ({diagInfo, player, expandedMode, showHelp}) => {

  const doSame4Analysis = (offOrDef: "off" | "def", dir: 1 | -1) => {
    const field = "adjEff";
    const contrib = (lineupPlusDiag: any, field: string) =>
      (lineupPlusDiag?.contrib?.[offOrDef]?.[field] || 0) as number;
    const lineups = diagInfo.filter(
      (lineupPlusDiag) => contrib(lineupPlusDiag, field)*dir < 0
    );
    const contribs = lineups.map((lineupPlusDiag) => contrib(lineupPlusDiag, field));
    const sumContrib = _.sum(contribs)
    const meanContrib: any = contribs.length ? mean(contribs) : 0;
    const stdevContrib: any = contribs.length ? std(contribs) : 0;
    return [
      _.chain(lineups).filter( // has to be > 1x stdev
        (lineupPlusDiag) => Math.abs(contrib(lineupPlusDiag, field) - meanContrib) > 1*stdevContrib
      ).filter( // only care about 1 direction
        (lineupPlusDiag) => (dir == -1) == (contrib(lineupPlusDiag, field) > meanContrib)
      ).sortBy(
        [ (lineupPlusDiag) => dir*contrib(lineupPlusDiag, field) ]
      ).take(5).map((lineupPlusDiag, i) => {
        return <li key={"" + i}>
          <a href="#">{lineupPlusDiag.keyArray.join(" / ")}</a>:
          &nbsp;contrib=[<b>{contrib(lineupPlusDiag, field).toFixed(2)}</b>]
          &nbsp;(poss=[<b>{contrib(lineupPlusDiag, "totalPoss").toFixed(1)}</b>], [<b>{(100*contrib(lineupPlusDiag, "possWeight")).toFixed(1)}%</b>])
        </li>
      }).value(),
      sumContrib, meanContrib, stdevContrib, contribs.length
    ];
  };
  const peerInfo = OnOffReportDiagUtils.buildPeerStatsForPlayer(player, diagInfo);
  const doPeerAnalysis = (offOrDef: "off" | "def") => {
    return _.keys(peerInfo).join(" , "); //TODO
  };


  const buildInfo = (offOrDef: "off" | "def") => {

    const infoHtml = _.flatMap([ -1, 1 ], (dir: 1 | -1, index) => {
      const [
        keyLineupHtml, keySum, keyMean, keyStdev, keyLength
      ] = doSame4Analysis(offOrDef, dir);

      const goodOrBad = (dir > 0) == (offOrDef == "off") ? "Bad" : "Good";

      return [
        <li key={offOrDef + index}>[<b>{keyLength}</b>] {goodOrBad} "Same-4"s, total contrib=[<b>{keySum.toFixed(1)}</b>] (mean=[<b>{keyMean.toFixed(1)}</b>], std=[<b>{keyStdev.toFixed(1)}</b>]). Key lineups:</li>,
        <ul key={"ul" + offOrDef + index}>
          {keyLineupHtml}
        </ul>,
        <li key={"peer" + offOrDef + index}>TODO {doPeerAnalysis(offOrDef)}</li>
      ]
    });

    return <ul>
      {infoHtml}
    </ul>;
  }

  return <span>
    {OnOffReportDiagUtils.getTitle(player, showHelp)}
    <ul>
      <li>Offensive analysis:</li>
      {buildInfo("off")}
      <li>Defensive analysis:</li>
      {buildInfo("def")}
    </ul>
    {expandedMode ?
      <span>For a more detailed analysis click here to view just this player.</span>
      : null
    }
  </span>;
}

export default RepOnOffDiagView;
