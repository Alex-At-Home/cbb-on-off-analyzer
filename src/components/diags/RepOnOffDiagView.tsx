// React imports:
import React, { useState } from 'react';

// Lodash
import _ from "lodash";

// Next imports:
import { NextPage } from 'next';

// Library Utils
// @ts-ignore
import { mean, std } from "mathjs";

// Utils
import { OnOffReportDiagUtils } from "../../utils/tables/OnOffReportDiagUtils";
import { CommonFilterParams } from "../../utils/FilterModels";
import { QueryUtils } from "../../utils/QueryUtils";

type Props = {
  diagInfo: Array<any>,
  player: Record<string, any>,
  playerMap: Record<string, string>, //(code -> id)
  commonParams: CommonFilterParams,
  expandedMode: boolean,
  onExpand: (playerId: string) => void,
  showHelp: boolean,
  keyLineupThreshold?: number //(for testing, leave blank in prod)
};

const RepOnOffDiagView: React.FunctionComponent<Props> = ({diagInfo, player, playerMap, commonParams, expandedMode, onExpand, showHelp, keyLineupThreshold}) => {

  const threshold = _.isNil(keyLineupThreshold) ? 1 : keyLineupThreshold;

  // Build player comparison links with peers:
  const baseMaybeAdvQuery = QueryUtils.extractAdvancedQuery(commonParams.baseQuery || "");
  const compareLinkFromPeer = (peerCode: string) => {
    return OnOffReportDiagUtils.buildPlayerComparisonLink(
      player.playerId, player.playerCode, playerMap[peerCode] || peerCode, peerCode, baseMaybeAdvQuery, commonParams
    );
  };

  const doSame4Analysis = (offOrDef: "off" | "def", dir: 1 | -1) => {
    const field = "adjEff";
    const contrib = (lineupPlusDiag: any, field: string) =>
      (lineupPlusDiag?.contrib?.[offOrDef]?.[field] || 0) as number;
    const lineups = diagInfo.filter(
      (lineupPlusDiag) => contrib(lineupPlusDiag, field)*dir < 0
    );
    const contribs = lineups.map((lineupPlusDiag) => contrib(lineupPlusDiag, field));
    const sumContrib = _.sum(contribs)
    const meanContrib: any = contribs.length ? mean(contribs) : 0;
    const stdevContrib: any = contribs.length ? std(contribs) : 0;
    const keyLineups =  _.chain(lineups).filter( // has to be > 1x stdev (unless threshold disabled)
        (lineupPlusDiag) => !threshold || (Math.abs(contrib(lineupPlusDiag, field) - meanContrib) > threshold*stdevContrib)
      ).filter( // only care about 1 direction
        (lineupPlusDiag) => (dir == -1) == (contrib(lineupPlusDiag, field) >= meanContrib)
      ).sortBy(
        [ (lineupPlusDiag) => dir*contrib(lineupPlusDiag, field) ]
      ).take(5).value();

    const keyPeers = _.chain(keyLineups).flatMap((v) => _.toPairs(v.peers)).transform((acc, kv) => {
        const peerId = kv[0] as string;
        const peerInfo = kv[1] as Record<string, any>;
        if (!acc.hasOwnProperty(peerId)) {
          acc[peerId] = peerInfo.poss as number;
        } else {
          acc[peerId] += peerInfo.poss as number;
        }
      }, {} as Record<string, number>).toPairs().sortBy([(kv) => -kv[1]]).take(5).value();

    return [
      keyLineups.map((lineupPlusDiag, i) => {
        const title = lineupPlusDiag.keyArray.join(" / ");
        const same4Ids = lineupPlusDiag.keyArray.map((code: string) => playerMap[code] || code);
        return <li key={"" + i}>
          {OnOffReportDiagUtils.buildOnOffAnalysisLink(player.playerId, same4Ids, commonParams, title)}
          &nbsp;contrib=[<b>{contrib(lineupPlusDiag, field).toFixed(2)}</b>]
          &nbsp;(poss=[<b>{contrib(lineupPlusDiag, "totalPoss").toFixed(1)}</b>], [<b>{(100*contrib(lineupPlusDiag, "possWeight")).toFixed(1)}%</b>])
        </li>
      }),
      keyPeers.map((peerKv, i) => {
        const append = (keyPeers.length - 1 == i) ? "" : ", ";
        return <span key={"peer" + i}>{compareLinkFromPeer(peerKv[0])} (poss=[<b>{peerKv[1].toFixed(0)}</b>]){append}</span>
      }),
      sumContrib, meanContrib, stdevContrib, contribs.length
    ];
  };

  const buildInfo = (offOrDef: "off" | "def") => {

    const infoHtml = _.flatMap([ -1, 1 ], (dir: 1 | -1, index) => {
      const [
        keyLineupHtml, keyPeerHtml, keySum, keyMean, keyStdev, keyLength
      ] = doSame4Analysis(offOrDef, dir);

      const goodOrBad = (dir > 0) == (offOrDef == "off") ? "Bad" : "Good";

      return [
        <li key={offOrDef + index}>[<b>{keyLength}</b>] {goodOrBad} "Same-4"s, total contrib=[<b>{keySum.toFixed(1)}</b>] (mean=[<b>{keyMean.toFixed(1)}</b>], std=[<b>{keyStdev.toFixed(1)}</b>]).
        {keyLineupHtml.length > 0 ? " Key lineups:" : " All lineups similar."}</li>,
        <ul key={"ul" + offOrDef + index}>
          {keyLineupHtml.length > 0 ? keyLineupHtml.concat(
            [<li key={"ul" + offOrDef + "peers"}>Key sub-ins: {keyPeerHtml}</li>]
          ) : null}
        </ul>
      ]
    });

    return <ul>
      {infoHtml}
    </ul>;
  }

  return <span>
    {OnOffReportDiagUtils.getTitle(player, showHelp)}
    <ul>
      <li>Offensive analysis ([<b>{player?.replacement?.off_adj_ppp?.value.toFixed(1)}</b>] pts/100):</li>
      {buildInfo("off")}
      <li>Defensive analysis ([<b>{player?.replacement?.def_adj_ppp?.value.toFixed(1)}</b>] pts/100):</li>
      {buildInfo("def")}
      {!expandedMode ?
        <li><a href="#" onClick={(event) => { event.preventDefault(); onExpand(player.playerCode) }}>To view more detailed analysis for just this player click here</a>.</li>
        :
        <li><a href="#" onClick={(event) => { event.preventDefault(); onExpand("") }}>To clear the detailed analysis and player filter click here</a>.</li>
      }
    </ul>
  </span>;
}

export default RepOnOffDiagView;
