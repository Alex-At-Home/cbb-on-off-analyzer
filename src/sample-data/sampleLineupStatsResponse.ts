export const sampleLineupStatsResponse = { //TODO: need to change this...
   "took": 204,
   "responses": [
      {
         "took": 204,
         "timed_out": false,
         "_shards": {
            "total": 1,
            "successful": 1,
            "skipped": 0,
            "failed": 0
         },
         "hits": {
            "total": {
               "value": 318,
               "relation": "eq"
            },
            "max_score": null,
            "hits": []
         },
         "aggregations": {
            "lineups": {
               "doc_count_error_upper_bound": 0,
               "sum_other_doc_count": 0,
               "buckets": [
                 {
                    "key": "AaWiggins_AnCowan_DaMorsell_ErAyala_JaSmith",
                    "doc_count": 38,
                    "total_off_fga": {
                       "value": 167
                    },
                    "total_def_2prim_attempts": {
                       "value": 40
                    },
                    "total_def_drb": {
                       "value": 68
                    },
                    "total_off_2pmid_made": {
                       "value": 12
                    },
                    "total_off_2pmid_attempts": {
                       "value": 44
                    },
                    "total_def_fta": {
                       "value": 47
                    },
                    "total_def_2p_made": {
                       "value": 40
                    },
                    "off_2pmidr": {
                       "value": 0.2634730538922156
                    },
                    "off_ppp": {
                       "value": 111.22448979591837
                    },
                    "off_3pr": {
                       "value": 0.38922155688622756
                    },
                    "total_off_3p_attempts": {
                       "value": 65
                    },
                    "total_off_2p_made": {
                       "value": 52
                    },
                    "total_def_fga": {
                       "value": 158
                    },
                    "total_def_2p_attempts": {
                       "value": 92
                    },
                    "def_poss": {
                       "value": 189
                    },
                    "total_off_2prim_made": {
                       "value": 40
                    },
                    "def_to": {
                       "value": 0.14285714285714285
                    },
                    "def_adj_opp": {
                       "value": 95.56989795918368
                    },
                    "off_2primr": {
                       "value": 0.3473053892215569
                    },
                    "def_2primr": {
                       "value": 0.25316455696202533
                    },
                    "total_off_pts": {
                       "value": 218
                    },
                    "total_off_orb": {
                       "value": 39
                    },
                    "off_adj_opp": {
                       "value": 103.93650793650794
                    },
                    "def_ppp": {
                       "value": 90.47619047619048
                    },
                    "total_def_2pmid_attempts": {
                       "value": 52
                    },
                    "def_2pmidr": {
                       "value": 0.3291139240506329
                    },
                    "total_off_3p_made": {
                       "value": 18
                    },
                    "def_3pr": {
                       "value": 0.4177215189873418
                    },
                    "total_def_3p_made": {
                       "value": 19
                    },
                    "total_def_pts": {
                       "value": 171
                    },
                    "total_def_orb": {
                       "value": 19
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 38,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "k5mWzG4B8OZ7gWZVgjRk",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DaMorsell",
                                         "id": "Morsell, Darryl"
                                      },
                                      {
                                         "code": "ErAyala",
                                         "id": "Ayala, Eric"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 102
                    },
                    "total_def_3p_attempts": {
                       "value": 66
                    },
                    "off_poss": {
                       "value": 196
                    },
                    "total_off_2prim_attempts": {
                       "value": 58
                    },
                    "total_def_2pmid_made": {
                       "value": 17
                    },
                    "total_def_2prim_made": {
                       "value": 23
                    },
                    "off_efg": {
                       "value": 0.47305389221556887
                    },
                    "off_to": {
                       "value": 0.14285714285714285
                    },
                    "total_off_drb": {
                       "value": 90
                    },
                    "total_off_fta": {
                       "value": 83
                    },
                    "def_efg": {
                       "value": 0.43354430379746833
                    },
                    "off_2p": {
                       "value": 0.5098039215686274
                    },
                    "def_2p": {
                       "value": 0.43478260869565216
                    },
                    "off_3p": {
                       "value": 0.27692307692307694
                    },
                    "def_3p": {
                       "value": 0.2878787878787879
                    },
                    "off_2prim": {
                       "value": 0.6896551724137931
                    },
                    "def_2prim": {
                       "value": 0.575
                    },
                    "off_2pmid": {
                       "value": 0.2727272727272727
                    },
                    "def_2pmid": {
                       "value": 0.3269230769230769
                    },
                    "off_ftr": {
                       "value": 0.49700598802395207
                    },
                    "def_ftr": {
                       "value": 0.2974683544303797
                    },
                    "off_orb": {
                       "value": 0.3644859813084112
                    },
                    "def_orb": {
                       "value": 0.1743119266055046
                    }
                 },
                 {
                    "key": "AaWiggins_AnCowan_DaMorsell_DoScott_JaSmith",
                    "doc_count": 27,
                    "total_off_fga": {
                       "value": 87
                    },
                    "total_def_2prim_attempts": {
                       "value": 27
                    },
                    "total_def_drb": {
                       "value": 43
                    },
                    "total_off_2pmid_made": {
                       "value": 4
                    },
                    "total_off_2pmid_attempts": {
                       "value": 27
                    },
                    "total_def_fta": {
                       "value": 24
                    },
                    "total_def_2p_made": {
                       "value": 22
                    },
                    "off_2pmidr": {
                       "value": 0.3103448275862069
                    },
                    "off_ppp": {
                       "value": 89.1891891891892
                    },
                    "off_3pr": {
                       "value": 0.3563218390804598
                    },
                    "total_off_3p_attempts": {
                       "value": 31
                    },
                    "total_off_2p_made": {
                       "value": 22
                    },
                    "total_def_fga": {
                       "value": 85
                    },
                    "total_def_2p_attempts": {
                       "value": 58
                    },
                    "def_poss": {
                       "value": 114
                    },
                    "total_off_2prim_made": {
                       "value": 18
                    },
                    "def_to": {
                       "value": 0.2719298245614035
                    },
                    "def_adj_opp": {
                       "value": 93.37927927927929
                    },
                    "off_2primr": {
                       "value": 0.3333333333333333
                    },
                    "def_2primr": {
                       "value": 0.3176470588235294
                    },
                    "total_off_pts": {
                       "value": 99
                    },
                    "total_off_orb": {
                       "value": 18
                    },
                    "off_adj_opp": {
                       "value": 106.32105263157895
                    },
                    "def_ppp": {
                       "value": 77.19298245614036
                    },
                    "total_def_2pmid_attempts": {
                       "value": 31
                    },
                    "def_2pmidr": {
                       "value": 0.36470588235294116
                    },
                    "total_off_3p_made": {
                       "value": 7
                    },
                    "def_3pr": {
                       "value": 0.3176470588235294
                    },
                    "total_def_3p_made": {
                       "value": 9
                    },
                    "total_def_pts": {
                       "value": 88
                    },
                    "total_def_orb": {
                       "value": 14
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 27,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "apmWzG4B8OZ7gWZVgTSi",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DaMorsell",
                                         "id": "Morsell, Darryl"
                                      },
                                      {
                                         "code": "DoScott",
                                         "id": "Scott, Donta"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 56
                    },
                    "total_def_3p_attempts": {
                       "value": 27
                    },
                    "off_poss": {
                       "value": 111
                    },
                    "total_off_2prim_attempts": {
                       "value": 29
                    },
                    "total_def_2pmid_made": {
                       "value": 8
                    },
                    "total_def_2prim_made": {
                       "value": 14
                    },
                    "off_efg": {
                       "value": 0.3735632183908046
                    },
                    "off_to": {
                       "value": 0.18018018018018017
                    },
                    "total_off_drb": {
                       "value": 43
                    },
                    "total_off_fta": {
                       "value": 48
                    },
                    "def_efg": {
                       "value": 0.4176470588235294
                    },
                    "off_2p": {
                       "value": 0.39285714285714285
                    },
                    "def_2p": {
                       "value": 0.3793103448275862
                    },
                    "off_3p": {
                       "value": 0.22580645161290322
                    },
                    "def_3p": {
                       "value": 0.3333333333333333
                    },
                    "off_2prim": {
                       "value": 0.6206896551724138
                    },
                    "def_2prim": {
                       "value": 0.5185185185185185
                    },
                    "off_2pmid": {
                       "value": 0.14814814814814814
                    },
                    "def_2pmid": {
                       "value": 0.25806451612903225
                    },
                    "off_ftr": {
                       "value": 0.5517241379310345
                    },
                    "def_ftr": {
                       "value": 0.2823529411764706
                    },
                    "off_orb": {
                       "value": 0.29508196721311475
                    },
                    "def_orb": {
                       "value": 0.24561403508771928
                    }
                 },
                 {
                    "key": "AaWiggins_AnCowan_DoScott_ErAyala_JaSmith",
                    "doc_count": 24,
                    "total_off_fga": {
                       "value": 96
                    },
                    "total_def_2prim_attempts": {
                       "value": 28
                    },
                    "total_def_drb": {
                       "value": 36
                    },
                    "total_off_2pmid_made": {
                       "value": 10
                    },
                    "total_off_2pmid_attempts": {
                       "value": 25
                    },
                    "total_def_fta": {
                       "value": 27
                    },
                    "total_def_2p_made": {
                       "value": 17
                    },
                    "off_2pmidr": {
                       "value": 0.2604166666666667
                    },
                    "off_ppp": {
                       "value": 109.80392156862744
                    },
                    "off_3pr": {
                       "value": 0.4166666666666667
                    },
                    "total_off_3p_attempts": {
                       "value": 40
                    },
                    "total_off_2p_made": {
                       "value": 26
                    },
                    "total_def_fga": {
                       "value": 93
                    },
                    "total_def_2p_attempts": {
                       "value": 50
                    },
                    "def_poss": {
                       "value": 97
                    },
                    "total_off_2prim_made": {
                       "value": 16
                    },
                    "def_to": {
                       "value": 0.21649484536082475
                    },
                    "def_adj_opp": {
                       "value": 93.46372549019607
                    },
                    "off_2primr": {
                       "value": 0.3229166666666667
                    },
                    "def_2primr": {
                       "value": 0.3010752688172043
                    },
                    "total_off_pts": {
                       "value": 112
                    },
                    "total_off_orb": {
                       "value": 21
                    },
                    "off_adj_opp": {
                       "value": 104.70927835051546
                    },
                    "def_ppp": {
                       "value": 80.41237113402062
                    },
                    "total_def_2pmid_attempts": {
                       "value": 22
                    },
                    "def_2pmidr": {
                       "value": 0.23655913978494625
                    },
                    "total_off_3p_made": {
                       "value": 15
                    },
                    "def_3pr": {
                       "value": 0.46236559139784944
                    },
                    "total_def_3p_made": {
                       "value": 10
                    },
                    "total_def_pts": {
                       "value": 78
                    },
                    "total_def_orb": {
                       "value": 27
                    },
                    "players_array": {
                       "hits": {
                          "total": {
                             "value": 24,
                             "relation": "eq"
                          },
                          "max_score": 4.760141,
                          "hits": [
                             {
                                "_index": "bigten_2019",
                                "_type": "_doc",
                                "_id": "FJmWzG4B8OZ7gWZVjTbp",
                                "_score": 4.760141,
                                "_source": {
                                   "players": [
                                      {
                                         "code": "AaWiggins",
                                         "id": "Wiggins, Aaron"
                                      },
                                      {
                                         "code": "AnCowan",
                                         "id": "Cowan, Anthony"
                                      },
                                      {
                                         "code": "DoScott",
                                         "id": "Scott, Donta"
                                      },
                                      {
                                         "code": "ErAyala",
                                         "id": "Ayala, Eric"
                                      },
                                      {
                                         "code": "JaSmith",
                                         "id": "Smith, Jalen"
                                      }
                                   ]
                                }
                             }
                          ]
                       }
                    },
                    "total_off_2p_attempts": {
                       "value": 56
                    },
                    "total_def_3p_attempts": {
                       "value": 43
                    },
                    "off_poss": {
                       "value": 102
                    },
                    "total_off_2prim_attempts": {
                       "value": 31
                    },
                    "total_def_2pmid_made": {
                       "value": 3
                    },
                    "total_def_2prim_made": {
                       "value": 14
                    },
                    "off_efg": {
                       "value": 0.5052083333333334
                    },
                    "off_to": {
                       "value": 0.20588235294117646
                    },
                    "total_off_drb": {
                       "value": 43
                    },
                    "total_off_fta": {
                       "value": 19
                    },
                    "def_efg": {
                       "value": 0.34408602150537637
                    },
                    "off_2p": {
                       "value": 0.4642857142857143
                    },
                    "def_2p": {
                       "value": 0.34
                    },
                    "off_3p": {
                       "value": 0.375
                    },
                    "def_3p": {
                       "value": 0.23255813953488372
                    },
                    "off_2prim": {
                       "value": 0.5161290322580645
                    },
                    "def_2prim": {
                       "value": 0.5
                    },
                    "off_2pmid": {
                       "value": 0.4
                    },
                    "def_2pmid": {
                       "value": 0.13636363636363635
                    },
                    "off_ftr": {
                       "value": 0.19791666666666666
                    },
                    "def_ftr": {
                       "value": 0.2903225806451613
                    },
                    "off_orb": {
                       "value": 0.3684210526315789
                    },
                    "def_orb": {
                       "value": 0.38571428571428573
                    }
                 }
              ]
            }
         }
    }
  ]
}
