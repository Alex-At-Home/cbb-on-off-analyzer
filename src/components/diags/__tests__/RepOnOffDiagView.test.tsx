import renderer from 'react-test-renderer';
import React from 'react';

import { OnOffReportDiagUtils } from '../../../utils/tables/OnOffReportDiagUtils';
import { LineupUtils } from "../../../utils/stats/LineupUtils";
import { sampleLineupStatsResponse } from "../../../sample-data/sampleLineupStatsResponse"

import RepOnOffDiagView from "../RepOnOffDiagView";
import { LineupStatSet } from '../../../utils/StatModels';

describe("RepOnOffDiagView", () => {

  const lineupReport = {
    lineups: sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets as LineupStatSet[],
    avgEff: 100.0,
    error_code: "test"
  };
  const tempTeamReport = LineupUtils.lineupToTeamReport(
    lineupReport, true, 100, 10
  );
  const playersWithAdjEff = tempTeamReport.players || [];
  const diagInfo = OnOffReportDiagUtils.getRepOnOffDiagInfo(playersWithAdjEff[2], 100);

  [ true, false ].forEach((expandedMode) => {
    const threshold = expandedMode ? 0 : undefined; //test both with default threshold and with 0
    test(`RepOnOffDiagView - should create snapshot (expanded mode=[${expandedMode}], threshold=[${threshold}])`, () => {
      const component = renderer.create(<RepOnOffDiagView
        diagInfo={diagInfo || []}
        player={playersWithAdjEff?.[0]}
        playerMap={{"AaWiggins": "Wiggins, Aaron"}}
        expandedMode={expandedMode}
        commonParams={{baseQuery: "test"}}
        onExpand={(playerId: string) => false}
        showHelp={true}
        keyLineupThreshold={threshold}
      />);

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
