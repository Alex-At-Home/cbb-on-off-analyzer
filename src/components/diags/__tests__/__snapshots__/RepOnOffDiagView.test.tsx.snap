// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepOnOffDiagView RepOnOffDiagView - should create snapshot (expanded mode=[false], threshold=[undefined]) 1`] = `
<span>
  <span>
    <b>
      Replacement 'On-Off' 
      
      Diagnostics For [
      Wiggins, Aaron
      ]
    </b>
     
    <a
      href="https://hoop-explorer.blogspot.com/2020/03/diagnostics-mode-for-replacement-onoff.html"
      target="_blank"
    >
      (?)
    </a>
  </span>
  <ul>
    <li>
      Offensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          1.4
        </b>
        ] (mean=[
        <b>
          1.4
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          -1.5
        </b>
        ] (mean=[
        <b>
          -1.5
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      Defensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          2.7
        </b>
        ] (mean=[
        <b>
          2.7
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          -1.0
        </b>
        ] (mean=[
        <b>
          -1.0
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      <a
        href="#"
        onClick={[Function]}
      >
        To view more detailed analysis for just this player click here
      </a>
      .
    </li>
  </ul>
</span>
`;

exports[`RepOnOffDiagView RepOnOffDiagView - should create snapshot (expanded mode=[true], threshold=[0]) 1`] = `
<span>
  <span>
    <b>
      Replacement 'On-Off' 
      
      Diagnostics For [
      Wiggins, Aaron
      ]
    </b>
     
    <a
      href="https://hoop-explorer.blogspot.com/2020/03/diagnostics-mode-for-replacement-onoff.html"
      target="_blank"
    >
      (?)
    </a>
  </span>
  <ul>
    <li>
      Offensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          1.4
        </b>
        ] (mean=[
        <b>
          1.4
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         Key lineups:
      </li>
      <ul>
        <li>
          <div>
            <a
              href="/?autoOffQuery=false&baseQuery=test&offQuery=NOT%20%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&onQuery=%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&"
              target="_blank"
            >
              AaWiggins / AnCowan / ErAyala / JaSmith
            </a>
          </div>
           contrib=[
          <b>
            1.37
          </b>
          ]  (poss=[
          <b>
            264.2
          </b>
          ], [
          <b>
            46.5
            %
          </b>
          ])
        </li>
        <li>
          Key sub-ins: 
          <span>
            <div>
              <a
                href="/TeamReport?baseQuery=%28%7B%22Wiggins%2C%20Aaron%22%3B%22DoScott%22%7D%3D1%29%20AND%20%28test%29&filter=AaWiggins&incRepOnOff=true&repOnOffDiagMode=20&"
                target="_blank"
              >
                DoScott
              </a>
            </div>
             (poss=[
            <b>
              197
            </b>
            ])
            
          </span>
        </li>
      </ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          -1.5
        </b>
        ] (mean=[
        <b>
          -1.5
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      Defensive analysis ([
      <b />
      ] pts/100):
    </li>
    <ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Bad
         "Same-4"s, total contrib=[
        <b>
          2.7
        </b>
        ] (mean=[
        <b>
          2.7
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         Key lineups:
      </li>
      <ul>
        <li>
          <div>
            <a
              href="/?autoOffQuery=false&baseQuery=test&offQuery=NOT%20%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&onQuery=%22Wiggins%2C%20Aaron%22%20AND%20%7B%22Wiggins%2C%20Aaron%22%3B%22AnCowan%22%3B%22ErAyala%22%3B%22JaSmith%22%7D%3D4&"
              target="_blank"
            >
              AaWiggins / AnCowan / ErAyala / JaSmith
            </a>
          </div>
           contrib=[
          <b>
            2.66
          </b>
          ]  (poss=[
          <b>
            257.8
          </b>
          ], [
          <b>
            46.0
            %
          </b>
          ])
        </li>
        <li>
          Key sub-ins: 
          <span>
            <div>
              <a
                href="/TeamReport?baseQuery=%28%7B%22Wiggins%2C%20Aaron%22%3B%22DoScott%22%7D%3D1%29%20AND%20%28test%29&filter=AaWiggins&incRepOnOff=true&repOnOffDiagMode=20&"
                target="_blank"
              >
                DoScott
              </a>
            </div>
             (poss=[
            <b>
              197
            </b>
            ])
            
          </span>
        </li>
      </ul>
      <li>
        [
        <b>
          1
        </b>
        ] 
        Good
         "Same-4"s, total contrib=[
        <b>
          -1.0
        </b>
        ] (mean=[
        <b>
          -1.0
        </b>
        ], std=[
        <b>
          0.0
        </b>
        ]).
         All lineups similar.
      </li>
      <ul />
    </ul>
    <li>
      <a
        href="#"
        onClick={[Function]}
      >
        To clear the detailed analysis and player filter click here
      </a>
      .
    </li>
  </ul>
</span>
`;
